frontend/ui/fts.lua

```diff
@@ -0,0 +1,259 @@
+-- up to 9.0072*10^9 seconds.
+-- A year has 365.25*24*3600 = 3.15576*10^7 s, so we can store up to 285 years (9.0072e9/3.15576r7) with µs precision.
+
+-- A TV_PRECISION of 1e6 will give us a µs precision.
+local FTS_PRECISION = 1e6
+
+
+-- We prefer CLOCK_MONOTONIC_COARSE if it's available and has a decent resolution,
+-- as we generally don't need nano/micro second precision,
+-- and it can be more than twice as fast as CLOCK_MONOTONIC/CLOCK_REALTIME/gettimeofday...
+local PREFERRED_MONOTONIC_CLOCKID = C.CLOCK_MONOTONIC
+-- Ditto for REALTIME (for :realtime_coarse only, :realtime uses gettimeofday ;)).
+local PREFERRED_REALTIME_CLOCKID = C.CLOCK_REALTIME
+-- CLOCK_BOOTTIME is only available on Linux 2.6.39+...
+local HAVE_BOOTTIME = false
+if ffi.os == "Linux" then
+    -- Unfortunately, it was only implemented in Linux 2.6.32, and we may run on older kernels than that...
+    -- So, just probe it to see if we can rely on it.
+    local probe_ts = ffi.new("struct timespec")
+    if C.clock_getres(C.CLOCK_MONOTONIC_COARSE, probe_ts) == 0 then
+        -- Now, it usually has a 1ms resolution on modern x86_64 systems,
+        -- but it only provides a 10ms resolution on all my armv7 devices :/.
+        if probe_ts.tv_sec == 0 and probe_ts.tv_nsec <= 1000000 then
+            PREFERRED_MONOTONIC_CLOCKID = C.CLOCK_MONOTONIC_COARSE
+        end
+    end
+    logger.dbg("fts: Preferred MONOTONIC clock source is", PREFERRED_MONOTONIC_CLOCKID == C.CLOCK_MONOTONIC_COARSE and "CLOCK_MONOTONIC_COARSE" or "CLOCK_MONOTONIC")
+    if C.clock_getres(C.CLOCK_REALTIME_COARSE, probe_ts) == 0 then
+        if probe_ts.tv_sec == 0 and probe_ts.tv_nsec <= 1000000 then
+            PREFERRED_REALTIME_CLOCKID = C.CLOCK_REALTIME_COARSE
+        end
+    end
+    logger.dbg("fts: Preferred REALTIME clock source is", PREFERRED_REALTIME_CLOCKID == C.CLOCK_REALTIME_COARSE and "CLOCK_REALTIME_COARSE" or "CLOCK_REALTIME")
+
+    if C.clock_getres(C.CLOCK_BOOTTIME, probe_ts) == 0 then
+        HAVE_BOOTTIME = true
+    end
+    logger.dbg("fts: BOOTTIME clock source is", HAVE_BOOTTIME and "supported" or "NOT supported")
+
+    probe_ts = nil --luacheck: ignore
+end
+
+--[[--
+Fixed point time. Maps to a POSIX struct timeval (<sys/time.h>).
+]]
+local fts = {}
+
+--[[--
+Returns an fts time based on the current wall clock time.
+(e.g., gettimeofday / clock_gettime(CLOCK_REALTIME).
+
+This is a simple wrapper around util.gettime() to get all the niceties of a fts.
+If you don't need sub-second precision, prefer os.time().
+Which means that, yes, this is a fancier POSIX Epoch ;).
+
+@usage
+    local fts = require("ui/fts")
+    local fts_start = fts.realtime()
+    -- Do some stuff.
+    -- You can add and substract fts times
+    local fts_duration = fts.realtime() - fts_start
+
+@treturn fts fixed point time
+]]
+function fts.realtime()
+    local sec, usec = util.gettime()
```

Comment:
Might be worth it to reimplement it internally here, the only advantage `util.gettime` had was Windows support, but the whole clock_gettime thing means we've pretty much thrown that out the wondow (ha!) here ;p.

Reply:
You mean:
```
function fts.realtime()
    local timespec = ffi.new("struct timespec")
    C.clock_gettime(C.CLOCK_REALTIME, timespec)

    -- TIMESPEC_TO_FTS
    return tonumber(timespec.tv_sec) * FTS_PRECISION +  math.floor(fts.touSec(tonumber(timespec.tv_nsec / 1000)))
end
```

Reply:
Yup!